common/src/interlocked_hl.c

```diff
@@ -142,22 +165,42 @@ IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForNotVal
             {
                 /* Codes_SRS_INTERLOCKED_HL_42_002: [ If the value at address is not equal to value, InterlockedHL_WaitForNotValue shall return INTERLOCKED_HL_OK. ]*/
                 result = INTERLOCKED_HL_OK;
-                break;
+                continue_eval = false;
             }
-
-            /* Codes_SRS_INTERLOCKED_HL_42_006: [ If the value at address matches, InterlockedHL_WaitForNotValue shall issue another call to wait_on_address. ]*/
-
-            /* Codes_SRS_INTERLOCKED_HL_42_003: [ If the value at address is equal to value, InterlockedHL_WaitForNotValue shall wait until the value at address changes in order to compare it again to value by using wait_on_address. ]*/
-            /* Codes_SRS_INTERLOCKED_HL_42_004: [ When waiting for the value at address to change, the milliseconds argument value shall be used as timeout. ]*/
-            if (wait_on_address(address, current_value, milliseconds) != WAIT_ON_ADDRESS_OK)
+            else
             {
-                LogError("failure in wait_on_address(address=%p, &current_value=%p, milliseconds=%" PRIu32 ")",
-                    address, &current_value, milliseconds);
-                /* Codes_SRS_INTERLOCKED_HL_42_007: [ If wait_on_address fails, InterlockedHL_WaitForNotValue shall fail and return INTERLOCKED_HL_ERROR. ]*/
-                result = INTERLOCKED_HL_ERROR;
-                break;
+                /* Codes_SRS_INTERLOCKED_HL_42_006: [ If the value at address matches, InterlockedHL_WaitForNotValue shall issue another call to wait_on_address. ]*/
+
+                /* Codes_SRS_INTERLOCKED_HL_42_003: [ If the value at address is equal to value, InterlockedHL_WaitForNotValue shall wait until the value at address changes in order to compare it again to value by using wait_on_address. ]*/
+                /* Codes_SRS_INTERLOCKED_HL_42_004: [ When waiting for the value at address to change, the milliseconds argument value shall be used as timeout. ]*/
+                WAIT_ON_ADDRESS_RESULT wait_result = wait_on_address(address, current_value, milliseconds);
+                switch (wait_result)
+                {
+                    case WAIT_ON_ADDRESS_OK:
+                    {
+                        result = INTERLOCKED_HL_OK;
```

Comment:
>result = INTERLOCKED_HL_OK; [](http://example.com/codeflow?start=24&length=27)

same comments as above for all the code.

wait_on_address can wake up for WHATEVER reason it wants (... tsk tsk tsk).

Assigning here "result" cannot be done, because a determination of the "real" value has not been performed yet (only line 163) does that "for real".